### ISSUE TITLE
Don't assume instance/class method by signature.

### ISSUE THREAD

**Issue by JustKiddingCode** [pydantic/pydantic#10177]
Some functions in pydantic rely on parameter names to determine if a method is a class or instance method.

https://github.com/pydantic/pydantic/blob/823d51f92f544614bca556076e27dbacd1298491/pydantic/_internal/_decorators.py#L676
```py
def is_instance_method_from_sig(function: AnyDecoratorCallable) -> bool:
    sig = signature(unwrap_wrapped_function(function))
    first = next(iter(sig.parameters.values()), None)
    if first and first.name == 'self':
        return True
    return False
```

Since using `self` and `cls` is only a convention (although coded in PEP-8), it shouldn't be used.
Using the inspect module, this can be improved:

```py
import inspect

if inspect.ismethod(cls.method) and cls.method.__self__ is cls:
    # method bound to the class, e.g. a classmethod
```

see https://stackoverflow.com/a/19228282 for a details.

**Comment by sydney-runkle** [comment]
My hunch is that we do this for performance reasons - happy to accept a PR with this alternative if we also add a benchmark showing that the new method isn't any slower!

**Comment by sydney-runkle** [comment]
@JustKiddingCode, any interest in pursuing this?

**Comment by JustKiddingCode** [comment]
Yes, I'd like to tackle this. Might take a month or two, before I can start.

**Comment by Rony-ZenAlden** [comment]
Hi — I'd like to pick this up, since the discussion went quiet about a year ago.

Before writing any code I want to confirm the intended direction, because the
classification happens at **decorator-application time — inside the class body,
before the function is bound to a class**. At that point `inspect.ismethod()` /
`__self__` aren't available, so binding-based detection (as in the original
suggestion) can't work there; beyond an explicit `@classmethod`/`@staticmethod`
wrapper, the first-parameter name is the only available signal.

Given that, which direction would you accept?

1. **Keep the heuristic, harden the edges** — raise a clear `PydanticUserError`
   when a plain function is ambiguous (first param is neither `self`/`cls` nor an
   explicit decorator), and handle positional-only / unsignaturable functions.
   Non-breaking.
2. **Deprecate name-based auto-`classmethod`** and require an explicit
   `@classmethod`, via a deprecation path. More correct long-term, but breaking.

Happy to implement whichever you'd accept — I'm leaning toward (1) as the
non-breaking option. cc @sydney-runkle
